### ISSUE TITLE
why is hotp not working with the token provided from google authenticator.

### ISSUE THREAD

**Issue by SiestaMadokaist** [yeojz/otplib#343]
**Describe the bug**
I tried to verify an hotp from my google auth, but it never success.

I generated my qrcode using this website: https://stefansundin.github.io/2fa-qr/
and my otp-uri is: 
`otpauth://hotp/admin?secret=ABCDEFGHIJKLMNOP&issuer=me&counter=2`;

the otp token I get from google auth is: `281014`
but it doesn't get verified

```javascript
const otpFromGoogleAuth = '281014'
const otplib = require('otplib');
const secret = 'ABCDEFGHIJKLMNOPQ';
const counter = 2;
const token = otplib.hotp.generate(secret, counter);
// >>> 083514
const verified = otplib.hotp.check(otpFromGoogleAuth, secret, counter);
// >>> false


**To Reproduce**
Steps to reproduce the behavior:
1. use google authenticator
2. create an hotp with the secret mentioned above
3. compare it with the hotp generated from otplib

**Expected behavior**
token from google auth should be the same with token from otplib.
or at least otplib.verify should return true, given the token from google auth.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Details (please provide any relevant information):**
```
> otplib.hotp.allOptions()
{
  algorithm: 'sha1',
  createHmacKey: [Function: hotpCreateHmacKey],
  createDigest: [Function: createDigest],
  digits: 6,
  encoding: 'ascii'
}
```

- otplib version: 12.0.1
- os: mac osx
- browser: -
- node / yarn: 12.8.0
- frameworks (if any): -

**Additional context**
Add any other context about the problem here.

did I used an incorrect config or something?

**Comment by yeojz** [comment]
acknowledged.

**Comment by benyaminl** [comment]
@yeojz 

Is there any specific times or by design the HOTP should not work with Google Authenticator?

**Comment by anu-sree-1** [comment]
> **Expected behavior**
> token from google auth should be the same with token from otplib.
> or at least otplib.verify should return true, given the token from google auth.

@yeojz
I'm going through a similar issue, but with totp.

- Google auth accepts secret strings via "Enter a Setup Key" option that are not strictly base32 too. Just that the string needs to be 16 characters long, and it must not include 8 or 9.
-  The same is not with otplib. It stictly checks for base32 format. Which means the secret needs to be longer than 16 characters, must not include number 1, etc. Because of this I'm not able to generate same totp generated in google auth using the same secret key

This issue had not been spotted previously because v12 otplib used to not striclty check for base32 strings when implemented in node (not in browser). Yet the code generated used to be different even then. But didnt compare with google auth and notice the difference then.

∘ secret I used for testing: "1234567123456712"

**Comment by yeojz** [comment]
@anu-sree-1 

**Regarding base32**

- Google Authenticator's keyuri spec does stipulate that it *has* to be base32 (https://github.com/google/google-authenticator/wiki/Key-Uri-Format). 

- I downloaded both ADR and IOS google's authenticator and attempted to key in `1234567123456712`, it actually doesn't let me save though.... May I know which app or version you are using?

---

**Regarding Support of arbitrary strings**

- If you are using `v13.x`, you can technically bypass base32 decoding or encoding by passing a `Unit8Array` instead of a `string` as secret. You can make use of the `stringToBytes` helper or do the conversion with your own code.

- I cannot guarantee behaviour as I haven't tested for random strings, but result should theoratically be stable.
  - **EDIT**: tried with `23456723456723456723456723456` and it can generate the same if I utilise `secret: stringToBytes('23456723456723456723456723456')`



- If you need shorter byte length, you can set a [custom guardrail ](https://otplib.yeojz.dev/guide/danger-zone)

---

<img src="https://github.com/user-attachments/assets/16607c56-66a8-4718-99c2-e73f0d250efa" width="250px" />

**Comment by anu-sree-1** [comment]
@yeojz
Thank you for the details

- Google Authenticator was of v6.  Once I updated it to v7, older secrets that were not strictly base32 still exists and is supported, but they have put strict base32 validation for newer secrets now.
- I tried using stringToBytes and then generating totp. It is still not matching the totps being generated on older non-base32 secrets by GAuth. But as for my code I think supporting normal strings is fair enough for now. Just wanted to mention it in case it could be looked into.

**Comment by yeojz** [comment]
@anu-sree-1 
perhaps if you got `1234567123456712` still in generating in authenticator, can give me a timestamp + code pair that is generated by your authenticator, so that I can see if there are discrepencies since I have no way to add to authenticator now..

Just to be sure, if you are testing locally and comparing, is your comparison code something like this?

```ts
const secret = stringToBytes("1234567123456712");

const guardrails = createGuardrails({
  MIN_SECRET_BYTES: 1,
});

const token = await generate({
  secret,
  strategy: "totp",
  guardrails,
});
```

**Comment by anu-sree-1** [comment]
@yeojz 

secret: 1234567123456712 
unix timestamp: 1768500360
totp generated by g-auth: 761777
totp generated by otplib auth: 702708

code I used:
```ts
import {
  generate,
  ScureBase32Plugin,
  createGuardrails,
} from "otplib";
import { stringToBytes } from "@otplib/core";

const base32 = new ScureBase32Plugin();
const totpOptions = { guardrails: createGuardrails({ MIN_SECRET_BYTES: 10 }) };

function isValidBase32(str: string) {
  try {
    base32.decode(str);
    return true;
  } catch (error) {
    return false;
  }
}

const secretB32 = isValidBase32(secret) ? secret: base32.encode(stringToBytes(secret));
const totp = await generate({ ...totpOptions, secret: secretB32 });
```

Reference to my complete implementation: [github link](https://github.com/anu-sree-1/totp-generation-app/blob/master/server-side-generated/server/index.ts)

---

Another pair of totp:
| otplib auth | g-auth |
|---|---|
|<img width="300" height="300" alt="Image" src="https://github.com/user-attachments/assets/31178c04-1e21-4d9b-905f-3bddc152424b" />|<img width="300" height="300" alt="Image2" src="https://github.com/user-attachments/assets/e4f2a982-8c99-4aa1-8d10-e151fe85b935"/>|

**Comment by yeojz** [comment]
@anu-sree-1 


Found the reason for the difference, results:
<img width="270" height="113" alt="Image" src="https://github.com/user-attachments/assets/f5655b17-c367-47e8-bf37-244e1f3816dc" />

**Root cause:**

Google Authenticator silently did some normalising of the string instead of rejecting outright in older versions.

You will need to normalise the character set to Base32 characters. So taking your example above:

```ts
function normaliseCharset(input) {
  return input
    .toUpperCase()
    .replace(/ /g, "")   // Remove spaces
    .replace(/1/g, "I")  // Replace '1' with 'I' (capital i)
    .replace(/0/g, "O"); // Replace '0' with 'O' (capital o)
}

const normalised = isValidBase32(secret) 
  ? secret
  : normaliseCharset('1234567123456712'); // EDITED.

const totp = await generate({ 
   ...totpOptions, 
  secret: normalised 
});
```

**Tip**: You do not need to encode the secret if you already converted it to bytes.


**Updated docs**: https://otplib.yeojz.dev/guide/troubleshooting.html#otp-not-matching-google-authenticator

**Comment by anu-sree-1** [comment]
@yeojz 

That normalisation helps!

But a small issue. It doesnt seem to work in combination with stringToBytes.

<img width="800" height="600" alt="Image" src="https://github.com/user-attachments/assets/f5cc2654-c6f7-42ff-9654-a3d9c23dcb0e" />


<br/><br/>
<strong>g-auth totp:  </strong>
<img width="300" height="200" alt="Image" src="https://github.com/user-attachments/assets/b724bcd9-b410-46b9-80ed-a1c24a260201" />

**Comment by yeojz** [comment]
@anu-sree-1 You're right! I went back to check and yes, I didn't use stringToBytes again.. I had went through the same thought route and probably mixed up typing the reply.. 😓 

<img width="434" height="609" alt="Image" src="https://github.com/user-attachments/assets/54430a88-4ff5-4709-93d7-9f0de133f148" />

This makes sense as Authenticator would have assumed everything was base32.
So I would assume when they receive a non-base32, they normalise then save it as is, which is then treated as a base32 string in future generation.

Note: EDITED the previous reply so if anyone copying in the future will be good to go.

**Comment by anu-sree-1** [comment]
Thanks for the clarification! 😊
